Network Viewer options

thiagoftsm opened the pull request:
##### Summary
Fixes #9455

##### Component Name
Collector

##### Test Plan

1 -  Compile Netdata with the option `-DNETDATA_INTERNAL_CHECKS`
2 - Configure your `ebpf.conf` inside `/etc/netdata/`, for example, use
```
[global]
    ebpf load mode = return
    #ebpf load mode = entry
    disable apps = no

[ebpf programs]
    process = yes
    network viewer = yes

[network viewer]
    maximum dimensions = 20
    ports = !145 !139 100-1024 144  ssh * https 3128:2000 60000-70000 78945 !compressnet iafservers
    hostnames = netdata.cloud netdata.io    !*example*
    ips = 127.0.0.1 192.168.0.0/6 172.20.0.0-172.20.2.255 4204:14c:5b72:1234:5678:1aee:ac11:cbb5/24 1200:10c:1234:5678:1ee2:1aee:ac11:cbb5/90 192.168.0.100-192.168.0.50 192.168.0.0/24     c704:14c:5b72:888d:1ee2:1aee:ac11:cbb5  9436:14c:1234:5678:1ee2:1aee:ac11:cbb0-9436:14c:1234:5678:1ee2:1aee:ac11:cbff 9436:14c:1234:5678:1ee2:1aee:ac11:cbf2 !*

[service name]
    19999 = NetdataTLS
    20000 = Netdata
```

3 - Start your Netdata
4 - Wait for few seconds and run the next command and confirm that the options was parsed and linked as expected:

```
$ grep "network viewer" /var/log/netdata/error.log ; grep "ignored" /var/log/netdata/error.log 
```

##### Additional Information

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

underhood commented on collectors/ebpf.plugin/ebpf.c:
```diff
@@ -633,6 +638,557 @@ static inline int parse_disable_apps(char *ptr)
+ * Parse the port ranges given and create Network Viewer Port Structure
+ *
+ * @param out is the output link list
+ * @param ptr is a pointer with the text to parser.
```
typo: with the text to ~parser~ -> parse

underhood commented on collectors/ebpf.plugin/ebpf.c:
```diff
@@ -633,6 +638,557 @@ static inline int parse_disable_apps(char *ptr)
+    if (co) {
+        struct config_option *cv;
+        for (cv = co->values; cv ; cv = cv->next) {
+            link_dimension_name(cv->name, cv->hash, cv->value);
+        }
+    }
+
+    //Always associated the default port to Netdata
+    ebpf_network_viewer_dim_name_t *names = network_viewer_opt.names;
+    uint16_t default_port = htons(19999);
+    for (; names->next; names = names->next) {
+        if (names->port == default_port) {
+            return;
+        }
+    }
+
+    link_dimension_name("Netdata", simple_hash("Netdata"), "19999");
```
in line `1176` you call `link_dimension_name(PORT, HASH, NAME);` but
in line `1189` you call `link_dimension_name(NAME, HASH, PORT);` ?

also is it good idea to have fixed 19999 ? what if user changes the port netdata listens on?

underhood replied:
`cv->name` here holds key name which in config as you described in PR description (and how I tested it) looks like this:

```
[service name]
    2222 = testing
```

then `cv->name` is the `port`. So I guess either configuration description is wrong or code. Or I misunderstand something?

thiagoftsm replied:
You are right, in the second call (line 1189), I inverted, because in normal situation I do not use numbers as name.

ilyam8 replied:
>   //Always associated the default port to Netdata

What purpose does it serve? Why we do it? 🤔

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

underhood commented on collectors/ebpf.plugin/ebpf.c:
```diff
@@ -633,6 +638,557 @@ static inline int parse_disable_apps(char *ptr)
+        }
+    }
+
+    link_dimension_name("19999", simple_hash("19999"), "Netdata");
```
thanks, now it is correct, however question still remains if hardcoding netdata to port 19999 is good idea. What if user runs netdata on another port?

thiagoftsm replied:
This is a good point, if the user runs in another port, the section `[service name]`needs being configured. For we read different ports, we would need to either read `netdata.conf` to get different values for `default port` and `bind to` or Netdata would need to pass these values as argument when the collector is called.

ilyam8 replied:
> For we read different ports, we would need to either read netdata.conf to get different values for default port and bind to or Netdata would need to pass these values as argument when the collector is called.

or we could expose env variable like `NETDATA_LISTEN_PORT` and use it here

thiagoftsm replied:
@ilyam8 I liked your suggestion, but this affects not only this plugin, but all plugins, so I think for we move forward with this, it would need another PR, because I will change the next function

https://github.com/netdata/netdata/blob/664d180dc7a5e34334a4f94399e5663eb5c0be8a/daemon/main.c#L672

ilyam8 replied:
> but all plugins

At the momenot it affects nothing, if you need this for this plugin it looks reasonable to add it in this PR. 

A PR with only new env var looks strange without context.

@mfundul @stelfrag @underhood what do u think guys?

underhood replied:
I think I agree with you @ilyam8 I worry for the confusion of users who use netdata on different ports

thiagoftsm replied:
All right, I created the environment variable to set the dimension when it is necessary.

underhood commented on collectors/ebpf.plugin/ebpf.c:
```diff
@@ -633,6 +638,557 @@ static inline int parse_disable_apps(char *ptr)
+ * @param out a pointer to store the link list
+ * @param range the informed range for the user.
+ */
+static void parse_port_list(void **out, char *range)
```
Behaviour here is inconsitent and I think this must be fixed:

1. If I put into list `-111` it will parse it as range `0-111` or by another words `-X`==`NETDATA_MINIMUM_PORT_VALUE-X`
2. But if I put into list `111-` it will parse it as single port `111`, to be consistent with point 1 it should parse as `X-NETDATA_MAXIMUM_PORT_VALUE`

So either such range should in both case give error or we should define it means until the port limit (min/max). (another option would be force it to be single port but I am strongly against this option).

Additionally if I put range `111-100` it will parse first to be `111` and last `100` there seems to be no check if `last>first`. In such case again either we should put error, or put warning and swap the values?

underhood replied:
additionally if I configure `included ports = 19999 - 12` the second parameter `-` will be parsed as range `0-0` this should be error instead.

ilyam8 replied:
> In such case again either we should put error, or put warning and swap the values?

i think warning is better, lets dont think for a user (`111-100` - it is likely he wanted `111-1000` and not `100-111`), if there is a mistake, log an error and that is it.

thiagoftsm replied:
@underhood when the first port is not specified I decided to ignore it, because we do not know exactly what the first value the user wants. We could assume the first possible value (`1`), but I am not sure if this is safe.
About the inverted order,  I am not discard the value, instead I am inverting it and writing a message inside the log.

underhood commented on collectors/ebpf.plugin/README.md:
````diff
@@ -191,6 +191,53 @@ The eBPF collector enables and runs the following eBPF programs by default:
+    included ips = 192.168.0.0/24 0001:0000:0000:0000:0000:0000:0000:0000/32 
+```
+
+When you define an `included` setting, Netdat will collect network metrics for that specific port. For example, if you
````
Netdat->Netdata,
port->port/service/hostname/ip?

underhood replied:
In general, even when I approve please make sure @joelhans takes a look at it before merging. There are multiple errors/strange formulations in the documentation. But I am not a native English speaker as well.

thiagoftsm replied:
I observed that when I copied and I pasted the format was cut, but the majority of the original text given by @joelhans  was kept. 
I fixed the typo that you requested. 
Now about the transition from `port` to `port/service/hostname/ip`, this would force to rewrite the whole paragraph. I understand that is enough to give an example for a pair of `included` and `excluded`, because all pairs works the same way.

vlvkobal replied:
> Now about the transition from port to port/service/hostname/ip...

Then you should write
```
When you define the `included ports` setting
```
And mention that other sections have similar behavior.

thiagoftsm replied:
I talked with @joelhans and we used @vlvkobal suggestion to update the paragraph. Thank you guys!

ilyam8 replied:
> 0001:0000:0000:0000:0000:0000:0000:0000/32

> https://en.wikipedia.org/wiki/IPv6_address#Representation

=> `1::/32`

Does your parser handle all cases?

> https://tools.ietf.org/html/rfc5952

Where did you get the ip address parser? Is it your custom? Are there any tests?

thiagoftsm replied:
> Does your parser handle all cases?

Yes, @ilyam8 the parser is done using `inet_pton` to do the conversion from string to network byte order, so all the values present in the RFC should be parsed in this function. 

> Where did you get the ip address parser?

The function `inet_pton` is defined inside `<arpa/inet.h>` and it is part of POSIX, you can see more details in this [link](https://pubs.opengroup.org/onlinepubs/9699919799/functions/inet_ntop.html)

>  Is it your custom? 

No, I never considered this possibility, because we would need to cover different formats for IPV4 and IPV6, so I decided to use the Linux functions.

> Are there any tests?

I compared the function output with IPV6 calculators around the internet, I also used `inet_ntop` to confirm that values converted from string to binary could be reversed from binary to string.

ilyam8 replied:
ok, thx

anyway `0001:0000:0000:0000:0000:0000:0000:0000/32` looks weird, nobody writes it like that

i suggest to use [`2001:DB8::/32`](https://tools.ietf.org/html/rfc3849) or whatever you like but written in _correct_ representation

thiagoftsm replied:
Thank you for the suggestion, I updated the documentation using the format you suggested.

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

ilyam8 commented on collectors/ebpf.plugin/ebpf.c:
```diff
@@ -681,36 +681,48 @@ static void parse_port_list(void **out, char *range)
 
-    w->last = (uint16_t)htons((uint16_t)test);
+    if (first > last) {
+        info("The specified order [%d, %d] is inverted, Netdata is fixing it!", first, last);
```
i dont think this is correct

example:

`333-100`

It is likely an error is `100` => should be `333-1000`. 

What u are doing here is creating a range based on your assimptions.

A user didnt ask to create `100-333` range.

ilyam8 replied:
Let user to fix it.

Ranges should reflect configuration file. At the moment they are not.

ilyam8 replied:
my point

`333-100`

The error likely:
- additional digit in the first, fixed: `33-100`
- missing digit in the last, fixed: `333-1000`


It is unlikely the user wants `100-333`

underhood replied:
Yeah I think ilyam is right every case of `-INT` `INT-` and reversed order is most likely misplaced space... maybe we should error in every case instead of trying to be "`smart`"

thiagoftsm replied:
I got your point of view, I pushed a commit that does not add the value to the list if the order is wrong.

ilyam8 commented on collectors/ebpf.plugin/ebpf.c:
```diff
@@ -712,10 +712,9 @@ static void parse_port_list(void **out, char *range)
-        int tmp = first;
-        first = last;
-        last = tmp;
+        info("The specified order %s is wrong, the smallest value is always the first!", copied);
```
do we have `warning`?

thiagoftsm replied:
No, we have `error`, `fatal` and `info`.

ilyam8 replied:
@mfundul i remember there was a talk about adding `warn`, i dont really understand why we dont still have it. Any ideas? What stops us? 😄

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

vlvkobal commented on collectors/ebpf.plugin/README.md:
```diff
@@ -191,6 +191,57 @@ The eBPF collector enables and runs the following eBPF programs by default:
 
+### `[network viewer]`
+
+You can configure the information shown on `outbund` and `inbound` charts with the settings in this section. 
```
outbund -> outbound

thiagoftsm replied:
Fixed with last commit.

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

joelhans commented on collectors/ebpf.plugin/README.md:
```diff
@@ -191,6 +191,57 @@ The eBPF collector enables and runs the following eBPF programs by default:
+-   `ips`: The IP or range of IPs that you want to monitor. You can use IPv4 or IPv6 addresses, use dashes to define a
+    range of IPs, or use CIDR values.
+    
+By default, Netdata displayes up to 500 dimensions on network viewer charts. If there are more possible dimensions, they
```
```suggestion
By default, Netdata displays up to 500 dimensions on network viewer charts. If there are more possible dimensions, they
```

thiagoftsm replied:
I fixed with last commit, thank you @joelhans !

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

vlvkobal commented:
```
[service name]
    19999 = NetdataTLS
```
```
2020-07-16 12:52:27:  INFO  : MAIN : Adding values NetdataTLS( 19999) to dimension name list used on network viewer
2020-07-16 12:52:27:  INFO  : MAIN : Adding values Netdata( 19999) to dimension name list used on network viewer
```
What's the point of adding two different names for the same port?

vlvkobal commented:
`excluded ips = 192.168.0.100-192.168.0.50`
```
2020-07-16 13:29:19:  INFO  : MAIN : Adding values 192.168.0.100 - 192.168.0.50 to IP list used on network viewer
```
Shouldn't it be ignored?

thiagoftsm commented:
> ```
> [service name]
>     19999 = NetdataTLS
> ```
> 
> ```
> 2020-07-16 12:52:27:  INFO  : MAIN : Adding values NetdataTLS( 19999) to dimension name list used on network viewer
> 2020-07-16 12:52:27:  INFO  : MAIN : Adding values Netdata( 19999) to dimension name list used on network viewer
> ```
> 
> What's the point of adding two different names for the same port?

I could not recreate this problem, so I rebase the PR and I push another commit. I am not sure if I miss to push a fix.

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

vlvkobal commented:
```
excluded ips = 192.168.0.100/-1
2020-07-16 13:42:11:  INFO  : MAIN : The specified CIDR -1 is not valid, the IP 192.168.0.100 will be ignored.
```
OK

```
excluded ips = 192.168.0.100/0
2020-07-16 13:43:30:  INFO  : MAIN : Adding values 192.168.0.100 - 255.255.255.255 to IP list used on network viewer
```
Wrong

```
excluded ips = 192.168.0.100/1
2020-07-16 13:44:19:  INFO  : MAIN : Adding values 192.168.0.100 - 255.255.255.255 to IP list used on network viewer
```
Wrong

```
excluded ips = 192.168.0.100/2
2020-07-16 13:45:17:  INFO  : MAIN : Adding values 192.168.0.100 - 255.255.255.255 to IP list used on network viewer
```
Wrong

```
excluded ips = 192.168.0.100/3
2020-07-16 13:45:41:  INFO  : MAIN : Adding values 192.168.0.100 - 223.255.255.255 to IP list used on network viewer
```
Wrong

vlvkobal commented:
`excluded ips = 192.168.0.100-192.168.0.50`
```
2020-07-16 13:56:29:  INFO  : MAIN : Adding values 192.168.0.100 - 192.168.0.50 to IP list used on network viewer
```

vlvkobal commented:
Ilya reminded me about tests and I agree that this code is a perfect candidate for unit tests. But, even more, it would be better to use a well tested third party library than writing our own code.

vlvkobal commented:
`included ips = 4422:3311::2266:1234-4422:3311::2266:1000`
```
2020-07-16 15:00:32:  INFO  : MAIN : Adding values 4422:3311::2266:1234 - 4422:3311::2266:1000 to IP list used on network viewer
```

thiagoftsm commented:
the last commit fixes these problems:

> 2020-07-16 12:52:27:  INFO  : MAIN : Adding values NetdataTLS( 19999) to dimension name list used on network viewer
> 2020-07-16 12:52:27:  INFO  : MAIN : Adding values Netdata( 19999) to dimension name list used on network viewer

and

> excluded ips = 192.168.0.100-192.168.0.50
> included ips = 4422:3311::2266:1234-4422:3311::2266:1000

It does not fix:

> excluded ips = 192.168.0.100/2

Because this is a problem that we inherit from the third library we are using, I am on it now.

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

ktsaou commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5810941&bid=19021977)

underhood commented on collectors/ebpf.plugin/ebpf.c:
```diff
@@ -653,6 +696,722 @@ static inline int parse_disable_apps(char *ptr)
+    if (likely(*end)) {
+        *end++ = '\0';
+        if (*end == '!') {
+            info("The exclusion cannot be in the second part of the range %s, it will be ignored.", copied);
```
I would make this more clear. Unclear for user if the exclusion will be ignore or the range.

thiagoftsm replied:
I updated the message saying that the whole group that is wrong will be ignored.

vlvkobal commented on collectors/ebpf.plugin/README.md:
```diff
@@ -191,6 +191,50 @@ The eBPF collector enables and runs the following eBPF programs by default:
+
+When you define a `ports` setting, Netdata will collect network metrics for that specific port. For example, if you
+write `ports = 19999`, Netdata will collect only connections for itself. The `ports`, `hostnames`, and `ips` settings
+all accept [simple patterns](/libnetdata/simple_pattern/README.md).
```
IPs and ports don't support simple patterns and they shouldn't. But we need to state it here somehow.

thiagoftsm replied:
I updated it, thanks to call attention

vlvkobal commented:
`ports = 100 50-150`

```
2020-07-21 12:06:25:  INFO  : MAIN : Adding values 100( 100, 100) to included port list used on network viewer
2020-07-21 12:06:25:  INFO  : MAIN : Adding values 50-150( 50, 150) to included port list used on network viewer
```

We are not ignoring ports if they are set before any range that includes them.
The same for IPs.

vlvkobal commented:
Is it a correct behavior when we depend on sequence of set values?

vlvkobal commented:
`ips = 192.168.0.123/2`
```
2020-07-21 12:16:19:  INFO  : MAIN : Adding values 192.168.0.123 - 255.255.255.255 to included IP list used on network viewer
```
Incorrect. It should be `192.0.0.0 - 255.255.255.255`.

vlvkobal commented:
`ips = 123:6432::2434/0`
```
2020-07-21 12:24:13:  INFO  : MAIN : Adding values 123:6432::2434 - 123:6432::ffff:ffff:ffff:ffff to included IP list used on network viewer
```

thiagoftsm commented:
> Is it a correct behavior when we depend on sequence of set values?

Initially I was thinking that this could be acceptable, but considering the stress tests I made, you are right. More comparison will consume more time, so I changed this in the last commit.

thiagoftsm commented:
> `ips = 192.168.0.123/2`
> 
> ```
> 2020-07-21 12:16:19:  INFO  : MAIN : Adding values 192.168.0.123 - 255.255.255.255 to included IP list used on network viewer
> ```
> 
> Incorrect. It should be `192.0.0.0 - 255.255.255.255`.

I also fixed this in the last commit. I was keeping the first address as the users inserted, because I thought somehow that could be the first value they really want, but considering that other IPV4 calculators update the network addresess, I updated the algorithm to use the correct network address.

thiagoftsm commented:
> `ips = 123:6432::2434/0`
> 
> ```
> 2020-07-21 12:24:13:  INFO  : MAIN : Adding values 123:6432::2434 - 123:6432::ffff:ffff:ffff:ffff to included IP list used on network viewer
> ```

This border case, that I fixed now, hat the same motive that I described before.
